staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go

```diff
@@ -264,22 +268,81 @@ func (c *webhookConverter) Convert(in runtime.Object, toGV schema.GroupVersion)
 	return converted, nil
 }
 
-func validateConvertedObject(unstructIn, unstructOut *unstructured.Unstructured) error {
-	if e, a := unstructIn.GetKind(), unstructOut.GetKind(); e != a {
+// validateConvertedObject checks that ObjectMeta fields match, with the exception of
+// labels and annotations.
+func validateConvertedObject(in, out *unstructured.Unstructured) error {
+	if e, a := in.GetKind(), out.GetKind(); e != a {
 		return fmt.Errorf("must have the same kind: %v != %v", e, a)
 	}
-	if e, a := unstructIn.GetName(), unstructOut.GetName(); e != a {
+	if e, a := in.GetName(), out.GetName(); e != a {
 		return fmt.Errorf("must have the same name: %v != %v", e, a)
 	}
-	if e, a := unstructIn.GetNamespace(), unstructOut.GetNamespace(); e != a {
+	if e, a := in.GetNamespace(), out.GetNamespace(); e != a {
 		return fmt.Errorf("must have the same namespace: %v != %v", e, a)
 	}
-	if e, a := unstructIn.GetUID(), unstructOut.GetUID(); e != a {
+	if e, a := in.GetUID(), out.GetUID(); e != a {
 		return fmt.Errorf("must have the same UID: %v != %v", e, a)
 	}
 	return nil
 }
 
+// deriveObjectMeta prepares out's metadata by copying that of in, but preserving out's labels and annotations.
+func deriveObjectMeta(in, out *unstructured.Unstructured)error {
+	obj := out.Object["metadata"]
+	convertedMetaData, ok := obj.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("invalid metadata of type %T in converted object", obj)
+	}
+
+	out.Object["metadata"] = in.Object["metadata"]
+
+	obj = out.Object["metadata"]
+	outMetaData, ok := obj.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("invalid metadata of type %T in input object", obj)
+	}
+
+	if obj, found := convertedMetaData["labels"]; !found || obj == nil {
+		delete(outMetaData, "labels")
+	} else {
+		convertedLabels, ok := obj.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("invalid metadata.labels of type %T in converted object", obj)
+		}
+		outLabels := map[string]interface{}{}
+		for k, v := range convertedLabels {
+			if v == nil {
+				v = ""
+			} else if _, ok := v.(string); !ok {
+				return fmt.Errorf("metadata.labels[%s] must be a string, but is %T in converted object", k, v)
+			}
+			outLabels[k] = v
+		}
+		outMetaData["labels"] = outLabels
+	}
+
+	if obj, found := convertedMetaData["annotations"]; !found || obj == nil {
```

Comment:
Can this routine be generalized into a function to avoid the near duplication of code between the labels and annotation handling? If split out might be a good candidate for unit tests.

Reply:
done